sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1244,6 +1244,23 @@ class Analyzer(
         case s @ ScalarSubquery(sub, conditions, exprId)
             if sub.resolved && conditions.isEmpty && sub.output.size != 1 =>
           failAnalysis(s"Scalar subquery must return only one column, but got ${sub.output.size}")
+        // SPARK-1854: block cases where GROUP BY columns
+        // are not part of the correlated columns
+        case s @ ScalarSubquery(sub, conditions, exprId)
+            if sub.resolved && sub.isInstanceOf[Aggregate] =>
```

Comment:
Nit (for a next time): It is easier to do this in the pattern match. For example:
`case s @ ScalarSubquery(agg: Aggregate, conditions, exprId) if agg.resolved => `